#!/bin/bash
# Usage: ./populateDB -u <user> -d <mysql_bin_directory>  -a <sandbox api server>
#  !!! The examples below work ONLY for mysql operating within MAMP !!!
# Example:
#  ./populateDB.sh -u gt_user -d /Applications/MAMP/Library/bin/ -a https://glygen.ccrc.uga.edu/sandbox/api
# Example:
# ./populateDB.sh -u gt_user -d /Applications/MAMP/Library/bin/ -a localhost:8080/api
#  This script populates the glycotree DB using the mysqlimport command
#  For security reasons, YOU must supply the 
#    DB user_name and directory containing mysqlimport
#    User will be prompted for the MYSQL password

# The commands in this script assume it is called from glycotree/SQL/
#  Thus, files specifying the model are in the directory ../model/

start=$(date)
echo $start
COUNT=0

# initiate (global) canonical_residues.csv with N_canonical_residues.csv
cp ../model/N_canonical_residues.csv ./canonical_residues.csv
# append canonical_residues.csv with O_canonical_residues.csv (remove header line)
awk 'NR > 1 {print($0);}' ../model/O_canonical_residues.csv >> ./canonical_residues.csv
cp ../model/enzymes.csv ./enzymes.csv
cp ../model/enzyme_mappings.csv ./enzyme_mappings.csv
cp ../model/rules.tsv ./rules.tsv
cp ../model/rule_data.tsv ./rule_data.tsv
	
# glycotree/SQL/compositions.csv is generated by glycotree/build_N-tree.sh
# glycotree/SQL/correlation.csv is generated by glycotree/build_N-tree.sh


if [ $# -lt 6 ]; then
  echo "Not enough arguments"
  exit
fi

for (( i=1; i<=$#; i++)); do
  j=$(($i + 1))

  if [ ${!i} = '-u' ]; then
    user=${!j}
    echo "user name $user"
  fi

  if [ ${!i} = '-d' ]; then
    dir=${!j}
    echo "mysql bin directory $dir"
  fi

  if [ ${!i} = '-a' ]; then
    api=${!j}
    echo "api site $api"
  fi
done

echo
echo "This macro will NOT put newly curated changes into the DB"
echo "The tsv file holding rule data MUST be updated to matcfh newly curated changes in the DB"
echo "Otherwise, deployment of the DB to the server will DESTROY the newly curated changes "
echo "Please review the the changes by visiting the Web page $api/curatorAdmin.html,"
echo "   then download, and execute the auto-generated macro"
read -p "Do you want to continue without reviewing curation data? (y/n)" goon
if ! [ $goon = 'y' ]; then
  echo "Point your browser to $api/curatorAdmin.html"
  exit
fi

echo "Continuing without reviewing curation data"

echo
read -s -p "Please enter the MYSQL password for $user: " pw
echo

echo
echo "populating SQL Tables"
echo

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="," --ignore-lines=1 --lines-terminated-by="\n" -c residue_name,residue_id,name,anomer,absolute,ring,parent_id,site,form_name,notes glycotree canonical_residues.csv 

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="," --ignore-lines=1 --lines-terminated-by="\n" -c type,orthology_group,uniprot,protein_refseq,dna_refseq,gene_name,gene_id,species,branch_site_specificity glycotree enzymes.csv 

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="," --ignore-lines=1 --lines-terminated-by="\n" -c glytoucan_ac,residue_name,residue_id,name,anomer,absolute,ring,parent_id,site,form_name,glycoct_index glycotree compositions.csv

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="," --ignore-lines=1 --lines-terminated-by="\n" -c glytoucan_ac,dp,homolog,relative_dp,shared glycotree correlation.csv
 
$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="," --ignore-lines=1 --lines-terminated-by="\n" -c residue_name,residue_id,type,uniprot,notes glycotree enzyme_mappings.csv

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="\t" --ignore-lines=1 --lines-terminated-by="\n" -c glytoucan_ac,base64_composition glycotree bitSet.tsv 

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="\t" --ignore-lines=1 --lines-terminated-by="\n" -c rule_id,class,description,logic glycotree rules.tsv 

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="\t" --ignore-lines=1 --lines-terminated-by="\n" -c instance,rule_id,focus,enzyme,other_residue,polymer,taxonomy,proposer_id,refs,comment,status,administrator,disputer_id glycotree rule_data.tsv 

$dir/mysqlimport -u $user -p$pw --local --delete -v --fields-terminated-by="\t" --ignore-lines=1 --lines-terminated-by="\n" -c id,name,auth,role glycotree curators.tsv 

$dir/mysqldump -u gt_user -p$pw glycotree > ./server/glycotree.sql
